### PR TITLE
Move projects properties to Directory.Build.props 

### PR DESCRIPTION
**Pull request by Shkarlatov**
Split *.csproj to local and shared properties.
It makes it easier to change global project properties and add a new conditions.

Documentation [Customize your build](https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019)

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    <RootNamespace>OpenRA</RootNamespace>
+    <OutputPath>../bin</OutputPath>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <ExternalConsole>false</ExternalConsole>
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
+    <CodeAnalysisRuleSet>..\OpenRA.ruleset</CodeAnalysisRuleSet>
+    <Configurations>Release;Debug</Configurations>
+
+    <Nullable>disable</Nullable>
```
Comment (penev92) [inline Directory.Build.props:16]: This seems to already be covered on line 50, where there is also a condition whether to disable?

**Comment by penev92** [comment]
Hey, thank you for your first contribution, it seems intriguing at a glance!

File: Directory.Build.props (hunk 1 above)
Comment (Shkarlatov) [reply Directory.Build.props:16]: Until LangVersion becomes 8.0, this will be disabled. I hope it will be used in the future

**Comment by pchote** [comment]
This looks really useful. The Windows .NET 5 build CI build is now consistently failing (i've rerun the workflow a couple of times to check), but i'm not sure if this is related to these changes or is some unrelated issue with the GH environment.

**Comment by Shkarlatov** [comment]
> This looks really useful. The Windows .NET 5 build CI build is now consistently failing (i've rerun the workflow a couple of times to check), but i'm not sure if this is related to these changes or is some unrelated issue with the GH environment.

May be add ```dotnet restore``` after ```dotnet nuget locals all --clear ``` in https://github.com/OpenRA/OpenRA/blob/fe129956bb1b3f7e3231c11f306f7387d36f0630/.github/workflows/ci.yaml#L65-L70

It github-env [issue](https://github.com/actions/virtual-environments/issues/1090)

**Comment by pchote** [comment]
Can you please try adding that in this PR?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -65,6 +65,8 @@ jobs:
         run: |
           # Work around runtime failures on the GH Actions runner
           dotnet nuget locals all --clear
+          dotnet nuget add source https://api.nuget.org/v3/index.json -n nuget.org
+          dotnet restore
```
Comment (pchote) [inline .github/workflows/ci.yaml]: Adding the source above appears to be the key fix. The restore should be run automatically as part of `dotnet build`, so isn't this line redundant?

Comment (Shkarlatov) [reply .github/workflows/ci.yaml]: Most likely. I remove this. Can I remove ``` dotnet nuget locals all --clear``` ?

Comment (pchote) [reply .github/workflows/ci.yaml]: I don't remember whether that was working around the same issue that the source fixes or something else. Try it, and see if the build fails?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -63,8 +63,7 @@ jobs:
       - name: Check Code
         shell: powershell
         run: |
-          # Work around runtime failures on the GH Actions runner
```
Comment (pchote) [inline .github/workflows/ci.yaml:66]: Can we please keep this comment (the nuget add source is still working around runtime failures after all), and squash the commits together?

File: OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj [hunk 4]
```diff
@@ -1,35 +1,10 @@
...
-  <PropertyGroup Condition="'$(Configuration)' == 'Debug'">
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
-    <Optimize>false</Optimize>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(TargetPlatform)' == 'win-x86'">
-    <Prefer32bit>true</Prefer32bit>
-  </PropertyGroup>
+
   <ItemGroup>
     <AdditionalFiles Include="../stylecop.json" />
```
Comment (teinarss) [inline OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]: Can be removed?

File: Directory.Build.props [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
+    <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
+    <RootNamespace>OpenRA</RootNamespace>
```
Comment (teinarss) [inline Directory.Build.props]: This causes some problems with the namespaces for all the projects that is not named `OpenRA`

File: Directory.Build.props [hunk 6]
```diff
@@ -0,0 +1,78 @@
+<Project>
```
Comment (teinarss) [inline Directory.Build.props:1]: There are formatting issues in this file.

File: OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj (hunk 4 above)
Comment (Shkarlatov) [reply OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]: Yes, we can remove Prefer32bit. Netcore ignoring this property [dotnet#5074](https://github.com/dotnet/project-system/issues/5074)

Comment (teinarss) [reply OpenRA.WindowsLauncher/OpenRA.WindowsLauncher.csproj]: I mean the line `<AdditionalFiles Include="../stylecop.json" />` which is already defined in the `Directory.Build.props `

File: Directory.Build.props (hunk 5 above)
Comment (Shkarlatov) [reply Directory.Build.props]: I watched how to build ra2 and thought all mods like this. But OpenSA use root directory of engine. Yeh, ```<RootNamespace>OpenRA</RootNamespace>``` need remove. (revert to OpenRA.Game.csproj ?)

File: Directory.Build.props (hunk 6 above)
Comment (Shkarlatov) [reply Directory.Build.props:1]: Will be fixed

File: Directory.Build.props (hunk 5 above)
Comment (teinarss) [reply Directory.Build.props]: Please remove this line

File: Directory.Build.props [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined in any of the csproj.

This appears to be a reasonable change to keep.

File: Directory.Build.props [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
Comment (pchote) [inline Directory.Build.props:5]: Note: This was defined on `OpenRA.Launcher`, `OpenRA.Mods.Common`, `OpenRA.Platform.Default` but not `OpenRA.Game`, `OpenRA.Mods.Cnc`, `OpenRA.Mods.D2k`, `OpenRA.Server`, `OpenRA.Test`, `OpenRA.Utility`, `OpenRA.WindowsLauncher`.

I'm not sure whether this changes anything we care about - we already have to copy all dependencies for packages releases, and we probably don't care that local builds include extra libraries?

File: Directory.Build.props [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
```
Comment (pchote) [inline Directory.Build.props]: Note: These two lines weren't defined on `OpenRA.Test` or `OpenRA.WindowsLauncher`.

These projects are never compiled with `Mono == true`, so this should be harmless.

File: Directory.Build.props [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
+    <OutputPath>../bin</OutputPath>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <ExternalConsole>false</ExternalConsole>
```
Comment (pchote) [inline Directory.Build.props:13]: Note: This wasn't defined on `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Platforms.Default`, `OpenRA.Test`

This will only be relevant for executable projects, so should be harmless on the other dlls.

File: Directory.Build.props [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
+    <CodeAnalysisRuleSet>..\OpenRA.ruleset</CodeAnalysisRuleSet>
+    <Configurations>Release;Debug</Configurations>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined on `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Platforms.Default`, `OpenRA.Test`

I have never found a definitive guide explaining msbuild's weird configurations system, so I don't know whether this matters or not. Is this even used with the latest csproj format? If yes, how does it interact with the configurations defined in the sln?

File: Directory.Build.props [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
+    <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined on `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Platforms.Default`, `OpenRA.Test`, `OpenRA.Utility`.

The docs are not helpful in explaining what binding redirects (which "override assembly unification") actually are, but this sounds like a feature we don't want. So probably harmless / wanted?

File: Directory.Build.props [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
+    <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined on `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Platforms.Default`, `OpenRA.Test`.

I can't find any documentation about what this does, but I imagine that if it is still used at all that it is only relevant for the executable projects. Probably harmless on the other dlls?

File: Directory.Build.props [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    <OutputPath>../bin</OutputPath>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <ExternalConsole>false</ExternalConsole>
+    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
+    <CodeAnalysisRuleSet>..\OpenRA.ruleset</CodeAnalysisRuleSet>
+    <Configurations>Release;Debug</Configurations>
+
+    <Nullable>disable</Nullable>
+    <AnalysisLevel>latest</AnalysisLevel>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined in any of the csproj

MSDN suggests this is already the default, so can be removed?

File: Directory.Build.props [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    <CodeAnalysisRuleSet>..\OpenRA.ruleset</CodeAnalysisRuleSet>
```
Comment (pchote) [inline Directory.Build.props]: This is a duplicate of line 21.

Please remove.

File: Directory.Build.props [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+    <SuppressNETCoreSdkPreviewMessage>true</SuppressNETCoreSdkPreviewMessage>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined in any of the csproj

This doesn't seem relevant to OpenRA, so remove?

File: Directory.Build.props [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    <SignAssembly>false</SignAssembly>
```
Comment (pchote) [inline Directory.Build.props]: Note: This wasn't defined in any of the csproj

I'm fairly certain that this must be the default, so remove?

File: Directory.Build.props [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <TargetPlatform Condition="$([MSBuild]::IsOsPlatform('Windows'))">win-x64</TargetPlatform>
```
Comment (pchote) [inline Directory.Build.props:25]: Note: These three weren't defined in `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Server`, `OpenRA.Test`, `OpenRA.Utility`, `OpenRA.WindowsLauncher`.

Probably harmless?

File: Directory.Build.props [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    <TargetPlatform Condition="$([MSBuild]::IsOsPlatform('Linux'))">linux-x64</TargetPlatform>
+    <TargetPlatform Condition="$([MSBuild]::IsOsPlatform('OSX'))">osx-x64</TargetPlatform>
+  </PropertyGroup>
+  
+  <PropertyGroup Condition="'$(TargetPlatform)' == 'win-x86'">
+    <Prefer32bit>true</Prefer32bit>
```
Comment (pchote) [inline Directory.Build.props]: Note: This was previously only defined on the executables.

Probably harmless?

File: Directory.Build.props [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+  </PropertyGroup> 
+  
+  <PropertyGroup Condition="'$(Configuration)' == 'Debug'">
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <Optimize>false</Optimize>
+  </PropertyGroup>
+    
+  <!-- In netstandard2.0, the BCL isn't annotated for nullability so we disable nullability there -->
```
Comment (pchote) [inline Directory.Build.props]: OpenRA won't work on anything below netstandard2.1, so this can be removed.

File: Directory.Build.props [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+  <Target Name="DisableAnalyzers" BeforeTargets="CoreCompile" Condition="'$(Configuration)'=='Release'">
+    <!-- Disable code style analysis on Release builds to improve compile-time performance -->
+    <ItemGroup Condition="'$(Configuration)'=='Release'">
+      <Analyzer Remove="@(Analyzer)" />
+    </ItemGroup>
+  </Target>
+  
+  <!-- StyleCop -->
+  <ItemGroup>
+      <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
```
Comment (pchote) [inline Directory.Build.props]: Note: These two weren't defined in `OpenRA.Launcher` or `OpenRA.WindowsLauncher`.

This appears to be a bugfix - These are simple projects that are hardly touched, so nobody noticed that style checking didn't work.

File: Directory.Build.props [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+      <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" />
+       <AdditionalFiles Include="../stylecop.json" />
+  </ItemGroup>
+  
+  <ItemGroup Condition="'$(Mono)' == ''">
```
Comment (pchote) [inline Directory.Build.props]: This block is only needed by `OpenRA.Game`. Can we please move it back?

File: Directory.Build.props [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <Optimize>true</Optimize>
+    <UseVSHostingProcess>false</UseVSHostingProcess>
+    <LangVersion>7.3</LangVersion>
+    <DebugSymbols>true</DebugSymbols>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <AutoGenerateBindingRedirects>false</AutoGenerateBindingRedirects>
+    <OutputPath>../bin</OutputPath>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
```
Comment (pchote) [inline Directory.Build.props]: Note: This is a duplicate of line 4.

Please remove.

**Review by pchote — COMMENTED** [review]
Here are my notes from comparing removals and additions. Most of these are just information, but there are a few change requests in there.

Once the fixups are applied I will tag a devtest build so we can verify that packaged releases have not regressed.

File: Directory.Build.props (hunk 11 above)
Comment (Shkarlatov) [reply Directory.Build.props]: It is still used. [A.Akinshin blog (BenchmarkDotNet maintainer)](https://github.com/AndreyAkinshin/aakinshin.net/blob/master/content/en/posts/2014/msbuild-configurations.md)

 >In certain cases, projects do not properly convey the Configuration so it is desirable to generate a solution with your own values. [slngen](https://microsoft.github.io/slngen/)

If not define **Configurations**, default build will be DEBUG.  Only if use `dotnet publish` or external use `dotnet build -c Release` will be build as Release

File: Directory.Build.props (hunk 7 above)
Comment (Shkarlatov) [reply Directory.Build.props]: [From docs](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#publish-related-properties)
> **AppendTargetFrameworkToOutputPath**
The AppendTargetFrameworkToOutputPath property controls whether the target framework moniker (TFM) is appended to the output path (which is defined by OutputPath). The .NET SDK automatically appends the target framework and, if present, the runtime identifier to the output path. Setting AppendTargetFrameworkToOutputPath to false prevents the TFM from being appended to the output path. However, without the TFM in the output path, multiple build artifacts may overwrite each other.
>
>For example, for a .NET 5.0 app, the output path changes from `bin\Debug\net5.0` to `bin\Debug`

>**AppendRuntimeIdentifierToOutputPath**
The AppendRuntimeIdentifierToOutputPath property controls whether the runtime identifier (RID) is appended to the output path. The .NET SDK automatically appends the target framework and, if present, the runtime identifier to the output path. Setting AppendRuntimeIdentifierToOutputPath to false prevents the RID from being appended to the output path.
>
>For example, for a .NET 5.0 app and an RID of win10-x64, the output path changes from `bin\Debug\net5.0\win10-x64` to` bin\Debug\net5.0 `

File: Directory.Build.props (hunk 13 above)
Comment (teinarss) [reply Directory.Build.props]: Should be, only used by the exe.

File: Directory.Build.props (hunk 8 above)
Comment (Shkarlatov) [reply Directory.Build.props:5]: [From docs](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#copylocallockfileassemblies)
>**CopyLocalLockFileAssemblies**
The CopyLocalLockFileAssemblies property is useful for plugin projects that have dependencies on other libraries. If you set this property to true, any NuGet package dependencies are copied to the output directory. That means you can use the output of dotnet build to run your plugin on any machine.

> Alternatively, you can use `dotnet publish` to publish the class library. For more information, see dotnet publish.

File: Directory.Build.props (hunk 12 above)
Comment (teinarss) [reply Directory.Build.props]: https://stackoverflow.com/a/29719064

File: Directory.Build.props (hunk 11 above)
Comment (teinarss) [reply Directory.Build.props]: > 
> 
> Note: This wasn't defined on `OpenRA.Mods.Cnc`, `OpenRA.Mods.Common`, `OpenRA.Mods.D2k`, `OpenRA.Platforms.Default`, `OpenRA.Test`
> 
> I have never found a definitive guide explaining msbuild's weird configurations system, so I don't know whether this matters or not. Is this even used with the latest csproj format? If yes, how does it interact with the configurations defined in the sln?

In the sln you define global build configurations that then also in the sln for each project you assign that global configuration to a project configuration.

Most of the times these maps 1:1.

```
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Any CPU = Debug|Any CPU
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{0DFB103F-2962-400F-8C6D-E2C28CCBA633}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{0DFB103F-2962-400F-8C6D-E2C28CCBA633}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{0DFB103F-2962-400F-8C6D-E2C28CCBA633}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{0DFB103F-2962-400F-8C6D-E2C28CCBA633}.Release|Any CPU.Build.0 = Release|Any CPU
```

File: Directory.Build.props (hunk 9 above)
Comment (teinarss) [reply Directory.Build.props]: It will cause a unsupported behavior on `OpenRA.Test` if running with Mono, but it will only be a problem if someone tries to build it directly as the test project is only built directly and not in the sln.

File: Directory.Build.props (hunk 13 above)
Comment (Shkarlatov) [reply Directory.Build.props]: It is still use. Visual Studio SDK 2015, 2017, 2019
> External components can access these properties through the Properties collection for the appropriate Visual Studio automation object. (That is, ProjectItem for project files and folders, Project for projects, and Configuration for configurations). This member name name can be used as an indexer into the Properties collection.

It was needed VS2015 debugger for edit&continue feature. Starting in Visual Studio 2017, the option to debug using the hosting process is no longer needed and has been removed.

Added for compatibility, but if targeting >=netstandard2.1 it no needed and must be remove.

File: Directory.Build.props (hunk 19 above)
Comment (Shkarlatov) [reply Directory.Build.props]: Maybe remove this? It is not supported by https://github.com/dotnet/project-system/issues/5074 and in netframework define use WoW64 or not.

File: Directory.Build.props (hunk 21 above)
Comment (Shkarlatov) [reply Directory.Build.props]: What should i do with it?

Comment (pchote) [reply Directory.Build.props]: I think we want to keep these, but will need to check the generated installers to be sure that this doesn't break anything.

File: Directory.Build.props (hunk 12 above)
Comment (Shkarlatov) [reply Directory.Build.props]: Then use self-deploy it not needed.

Comment (Shkarlatov) [reply Directory.Build.props]: I remove this. Will use default value.

**Comment by pchote** [comment]
Can you please now squash the fixup commits? https://github.com/OpenRA/OpenRA/wiki/Contributing#updating-a-pull-request

**Comment by pchote** [comment]
I pushed a test build to https://github.com/pchote/OpenRA/releases/tag/devtest-20210408

The macOS, AppImage, and Windows 32bit builds appear to work as expected. The Windows 64bit build doesn't launch and doesn't even appear to crash, no output at all. Can anyone else repro?

**Comment by Shkarlatov** [comment]
@pchote I tested builds. I cant build `dotnet build -r win-x86` (win-x64 also).  Without RID build is success. It is dotnet error https://github.com/dotnet/sdk/issues/14281 
Workaround: `dotnet build -r win-x64 -p:ImportByWildcardBeforeSolution=false`

**Comment by pchote** [comment]
That's odd. The x64 .NET 5 builds were working fine every time we've tested them before now - is something specific in this PR triggering the bug?

**Comment by Shkarlatov** [comment]
@pchote How build x64? I download OpenRA-release-20210321.zip (src) , run `dotnet build -r win-x86`.
> error NETSDK1134: Building a solution with a specific RuntimeIdentifier is not supported. If you would like to publish for a single RID, specifiy the RID at the individual project level instead. 

>C:\Program Files\dotnet\sdk\5.0.201\Current\SolutionFile\ImportAfter\Microsoft.NET.Sdk.Solution.targets(27,5): error NETSDK1134: сборка решения с заданным идентификатором RuntimeIdentifier не поддерживается. Если вы хотите выполнить публикацию для одного RID, укажите RID на уровне отдельного проекта. [D:\Downloads\OpenRA-release-20210321\OpenRA.sln]

Not sure if PR not contain problem.

**Comment by Shkarlatov** [comment]
I found how build and packaging. `msbuild -verbosity:m -nologo -t:Build -restore -p:Configuration=Release -p:TargetPlatform=win-x64` build success

**Comment by Shkarlatov** [comment]
```
devenv /DebugExe "D:\Downloads\OpenRA-devtest-20210408-x64-winportable\Dune2000.exe"
```
```
Exception thrown at 0x00007FFEFEA6D759 (KernelBase.dll) in Dune2000.exe: 0x04242420 (parameters: 0x0000000031415927, 0x00007FFEBDDD0000, 0x0000006B2A97E3B0).
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: HRException at memory location 0x0000006B2A97E778.
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: HRException at memory location 0x0000006B2A97E778.
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000006B2A97A1C8.
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: [rethrow] at memory location 0x0000000000000000.
Exception thrown at 0x00007FFEFEA6D759 in Dune2000.exe: Microsoft C++ exception: EEFileLoadException at memory location 0x0000006B2A97A1C8.
Failed to create CoreCLR, HRESULT: 0x80004005Failed to create CoreCLR, HRESULT: 0x80004005
```

File: Directory.Build.props [hunk 24]
```diff
@@ -0,0 +1,52 @@
+<Project>
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
```
Comment (pchote) [inline Directory.Build.props]: This is what breaks the x64 build. The 32 bit build is created first and some of its intermediate state is probably leaking into the 64 bit build.

Lets just remove this line to bring things back to how they currently are on bleed.

**Review by pchote — DISMISSED** [review]
LGTM and tested:
* All release packages now work as expected (https://github.com/pchote/OpenRA/releases/tag/devtest-20210411)
* Compiling using the Makefile works with both .NET5 and Mono
* Building and running through Rider still works as expected

For some reason debugging through Rider is not working on my system (hangs during startup), but this is also a problem on bleed so does not appear to be related to this PR.

**Comment by pchote** [comment]
It would be good if we could get this merged sooner rather than later for the CI fix.

File: Directory.Build.props [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    <TargetPlatform Condition="$([MSBuild]::IsOsPlatform('Linux'))">linux-x64</TargetPlatform>
+    <TargetPlatform Condition="$([MSBuild]::IsOsPlatform('OSX'))">osx-x64</TargetPlatform>
+  </PropertyGroup>
+  
+  <PropertyGroup Condition="'$(Configuration)' == 'Debug'">
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <Optimize>false</Optimize>
+  </PropertyGroup>
+    
+
```
Comment (penev92) [inline Directory.Build.props:34]: There are two empty lines here.

File: Directory.Build.props [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    <ItemGroup Condition="'$(Configuration)'=='Release'">
+      <Analyzer Remove="@(Analyzer)" />
+    </ItemGroup>
+  </Target>
+  
+  <!-- StyleCop -->
+  <ItemGroup>
+      <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
+      <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" />
+       <AdditionalFiles Include="../stylecop.json" />
```
Comment (penev92) [inline Directory.Build.props]: Indentation here is off.

File: Directory.Build.props [hunk 27]
```diff
@@ -0,0 +1,51 @@
+<Project>
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <TargetFramework Condition="'$(Mono)' == ''">net5.0</TargetFramework>
+    <TargetFramework Condition="'$(Mono)' != ''">netstandard2.1</TargetFramework>
```
Comment (penev92) [inline Directory.Build.props]: Can you pull out the two `TargetFramework` lines to a separate `PropertyGroup` like `TargetPlatform`? Also maybe the `<PlatformTarget>AnyCPU</PlatformTarget>` line.

I also am not clear what the difference between `TargetPlatform` and `PlatformTarget` is, if anyone can explain?

**Review by penev92 — DISMISSED** [review]
LGTM, just needs a few things to be addressed.

- [x] Also I notice `UseVSHostingProcess` properties have disappeared from 3 projects, but something tells me that's for the better?
- [x] And `AutoGenerateBindingRedirects` properties have also disappeared?

**Comment by pchote** [comment]
@penev92 see https://github.com/OpenRA/OpenRA/pull/19325#discussion_r609077552 and https://github.com/OpenRA/OpenRA/pull/19325#discussion_r609079822

**Comment by penev92** [comment]
Thanks.

**Comment by pchote** [comment]
I have force pushed fixups for @penev92's requests so we can merge this ASAP

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f3ab714041b9c5cfd1a767aed570ad1893f9367a...8c65199c897b61278abdc902ba07dbb250d27648
